tools/hook-hotplug

```diff
@@ -2,21 +2,31 @@
 # This file is part of cloud-init. See LICENSE file for license information.
 
 # This script checks if cloud-init has hotplug hooked and if
-# cloud-init has finished; if so invoke cloud-init hotplug-hook
+# cloud-init is ready; if so invoke cloud-init hotplug-hook
 
-is_finished() {
-    [ -e /run/cloud-init/result.json ]
+fifo=/run/cloud-init/hook-hotplug-cmd
+
+should_run() {
+    if [ -d /run/systemd ]; then
+        # check that the socket is ready
+        [ -p $fifo ]
+    else
+        # on non-systemd, check cloud-init fully finished.
+        [ -e /run/cloud-init/result.json ]
+    fi
 }
 
-if is_finished; then
-    # open cloud-init's hotplug-hook fifo rw
-    exec 3<>/run/cloud-init/hook-hotplug-cmd
-    env_params=" \
-        --subsystem=${SUBSYSTEM} \
-        handle \
-        --devpath=${DEVPATH} \
-        --udevaction=${ACTION} \
-    "
-    # write params to cloud-init's hotplug-hook fifo
-    echo "${env_params}" >&3
+if ! should_run; then
+    exit 0
 fi
+
+# open cloud-init's hotplug-hook fifo rw
+exec 3<>$fifo
+env_params=" \
+--subsystem=${SUBSYSTEM} \
```

Comment:
nit: Could we indent this and the following lines evaluate to a single string? It looks like you did this to avoid the extra indentation between the args (which I appreciate).  I think we can have the best of both worlds by quoting the strings on each line so that they get concatenated together before assignment.

Reply:
Do you mean something like:

```sh
env_params=""\
   " --subsystem=${SUBSYSTEM}"\
   " handle"\
   " --devpath=${DEVPATH}"\
   " --udevaction=${ACTION}"
```

This would be nice, but it is not POSIX compatible, it fails in dash and probably on other shells than bash.

I would just KISS and define this var in single line like the following, yes it has 90 characters, but it is more portable and readable IMO.

```sh
env_params=" --subsystem=${SUBSYSTEM} handle --devpath=${DEVPATH} --udevaction=${ACTION}"
```

Or we could use the short version of some parameter if 90 is not acceptable:

```sh
env_params=" -s=${SUBSYSTEM} handle --devpath=${DEVPATH} --udevaction=${ACTION}"
```

Reply:
> This would be nice, but it is not POSIX compatible, it fails in dash and probably on other shells than bash.

Oops, sorry for the bad suggestion.

> I would just KISS and define this var in single line like the following, yes it has 90 characters, but it is more portable and readable IMO.

+1, I'm fine with that

Reply:
Updated to single line.